File: epi_judge_python/buy_and_sell_stock_twice.py
from typing import List

from test_framework import generic_test


def buy_and_sell_stock_twice(prices: List[float]) -> float:
    diff = []

    result = 0.0

    cur_min = prices[0]
    for price in prices:
        if price < cur_min:
            cur_min = price
        result = max(result, price - cur_min)
        diff.append(result)

    cur_max = prices[-1]
    for i in range(len(prices) - 1, -1, -1):
        if prices[i] > cur_max:
            cur_max = prices[i]
        result = max(result, cur_max - prices[i] + diff[i])

    return result


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('buy_and_sell_stock_twice.py',
                                       'buy_and_sell_stock_twice.tsv',
                                       buy_and_sell_stock_twice))

File: epi_judge_cpp/buy_and_sell_stock_twice.cc
#include <vector>

#include "test_framework/generic_test.h"

using namespace std;

double BuyAndSellStockTwice(const vector<double> &prices) {
    vector<double> best_first_buy;
    double min_price_so_far = prices[0];
    double curr_best = 0;
    for (double price : prices) {
        curr_best = max(curr_best, price - min_price_so_far);
        min_price_so_far = min(min_price_so_far, price);
        best_first_buy.push_back(curr_best);
    }

    double result = curr_best;

    double max_price_so_far = prices.back();
    curr_best = 0;
    vector<double> best_second_buy(prices.size());
    for (int i = prices.size() - 1; i >= 0; i--) {
        curr_best = max(curr_best, max_price_so_far - prices[i]);
        max_price_so_far = max(max_price_so_far, prices[i]);
        best_second_buy[i] = curr_best;
    }

    for (int i = 1; i < prices.size() - 1; i++) {
        result = max(result, best_first_buy[i] + best_second_buy[i + 1]);
    }

    return result;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"prices"};
    return GenericTestMain(args, "buy_and_sell_stock_twice.cc",
                           "buy_and_sell_stock_twice.tsv", &BuyAndSellStockTwice,
                           DefaultComparator{}, param_names);
}
